### PR TITLE
perf(codegen): representation-selection Phase 3a — canonical string locals (tagged-at-rest Str rep)

### PR DESCRIPTION
**Pull request by proggeramlug**
Phase 3a of `docs/representation-selection-rfc.md` (§4 Str row, §5.5 boundary contracts, §5.6 GC rules): **canonical string locals, tagged-at-rest**. Builds on Phase 1 (#6903, `SlotRep`) and Phase 2 (#6905).

## Design: tagged-at-rest

`SlotRep::Str` marks a function local proven to hold NaN-box **string bits** — `STRING_TAG|ptr` for heap strings or inline `SHORT_STRING_TAG` SSO bits — at rest. Unlike canonical-i32, the rep does **not** move storage: the local keeps its `ctx.locals` double slot, its `js_shadow_slot_bind` GC binding (the mark/evacuation path is NaN-box-driven and handles these bits unchanged — **zero GC changes**), and every alias/refcount demote site keeps firing. SSO values stay by-value at rest (RFC §4). What the rep buys is a compile-time proof the string-op lowerings consume to **tag-dispatch inline** instead of routing operands through the opaque `js_get_string_pointer_unified` (which heap-materializes every SSO unbox and silently number-coerces).

## What gets deleted (per lowering)

| Site | Before | After |
|---|---|---|
| `s += part` (fn-local accumulator) | 2× opaque unified calls per iteration | 3-arm: both-heap → raw `js_string_append(h,h)` (in-place refcount==1 path kept) / both-string-SSO → `js_string_concat_box` / else → exact legacy sequence |
| `.length`, statically-string receiver | ~18-op GC-type-byte + forwarding tower | SSO → inline `lshr 40; and 0xFF`; heap → bare `load i32` of `utf16_len`; else → same `js_value_length_f64` slow call |
| `===`/`<` with a canonical-Str operand | 2 unified calls + compare | both-heap → direct `js_string_equals`/`js_string_compare` on raw handles; else → one SSO-aware call (`js_jsvalue_equals` / new `js_string_compare_value`) — never number-coercing |
| `charCodeAt`/`at`/`codePointAt` | unified call per receiver | proven-heap receiver → bare `and POINTER_MASK`; SSO/lie → legacy cold arm |
| `"lit" + value` coerce-concat | unified call on the literal side | inline `bitcast; and` (literals are proven-heap: pool handles from `js_string_from_bytes`) |
| `StringRef` materialization | `js_nanbox_string` call | inline `or STRING_TAG; bitcast`, null cold arm preserves the allocate-empty contract |

## Eligibility (under-approximate, mirror of Phase 1)

Selected at the `Stmt::Let` site for `String`/`StringLiteral`-typed locals in sync function bodies. Excluded: closure-referenced, boxed vars, module globals (incl. all entry/top-level code — same init-split constraints as Phase 1), async/generator/`was_plain_async`, catch bindings, locals with a non-string-proven reassignment, and locals equality-compared against a non-proven-string operand (the `compare.rs` `other_side_is_any` hazard) — via a ctx-free pre-pass (`collect_canonical_str_ineligible_locals`).

## Correctness (RFC §5.5 acceptance contract)

Every specialized site keeps a fallback arm: the hot arms fire only on runtime-tag-verified string bits, and anything else (a lying `string` annotation) routes to the exact pre-phase sequence or a strictly-more-correct non-coercing helper — never a new silent coercion. The 6 alias-demote sites (`let b = a`, scalar array/object stores, write-barrier choke point) are untouched and keep firing, so the in-place `+=` fast path stays alias-sound.

- **Gap test** `test-files/test_gap_repsel_canonical_str_locals.ts`: alias `+=` discipline (local, scalar-array, object-field), SSO round-trip (`+=`/`.length`/`===`/relational at ≤5 bytes, JSON-parsed SSO vs heap literal), lying-annotation acceptance, non-ASCII (2-byte UTF-8, CJK, surrogate-pair emoji) byte-exactness, route-compare and char-scan shapes. Verified byte-exact vs node in all four arms: flag on / flag off / each × `PERRY_GC_FORCE_EVACUATE=1`.
- `shadow_slot_hygiene.rs`: new test proves a canonical-Str local keeps its shadow-slot binding, the `+=` heap arm calls `js_string_append` with no unified call, and `.length` takes the tag dispatch, not the generic tower.
- Runtime unit test for `js_string_compare_value` (heap×SSO mixes, UTF-16 code-unit order, number-coercion parity with the legacy arm, invalid ranking).
- `PERRY_CANONICAL_STR_LOCALS` (default on) keyed into the object cache + key test.

## Structural proof (post-`opt -O3`, before → after)

Function-scoped kernel exercising the three recon shapes (`+=` accumulator / route-compare / char-scan), per-function `js_get_string_pointer_unified` call counts:

| shape | flag-off | flag-on |
|---|---|---|
| `+=` accumulator (3 sites, loop body) | 6 per iteration | **0 on the hot path** (9 remain confined to the `strapp.rcold`/`strapp.cold` fallback arms) |
| route-compare (`===` + `<` vs literals) | 4 | **0 anywhere in the function** — `streq.heap` calls `js_string_equals` directly, `strcmp.heap` calls `js_string_compare` directly |
| char-scan (`.length` + `charCodeAt` loop) | 2 | 0 in the loop (1 entry-once alias demote — the pre-existing discipline — plus 1 cold-arm fallback); `strlen.heap`/`strrecv.heap` arms contain **zero calls** (bare `load i32` / bare `and`-mask) |

Block-level audit of the emitted IR confirms the hot arms contain only the intended direct helper (`js_string_append` / `js_string_equals` / `js_string_compare` / no call at all for `.length` and the receiver mask); every remaining unified call sits in a cold fallback arm or an entry-once demote. App-pattern kernels (`string_concat_csv.ts`, `string_template_interp.ts`): module-wide unified calls 1 → **0**, with `rows[i].length` the 3-arm tag dispatch (heap arm = bare `load i32`).

A review pass hardened the self-append arm layout: a heap destination ALWAYS reaches `js_string_append` (so an SSO rhs cannot break the refcount==1 in-place accumulator into O(n²) copies), and the SSO-concat arm requires string tags on BOTH sides (a lying rhs keeps the legacy ToString coercion — `"ab" += 42` stays `"ab42"`).

**Wall-clock benchmarks deferred** — the bench box was unreachable at submission; timing follows as a PR comment when it returns.

## Deferred (reported residuals)

- Top-level/entry-context locals (module-global storage — same exclusion as Phase 1), so `bench_string_heavy`'s top-level `+=` loop keeps the legacy shape.
- The ~40 other string-method receiver sites (`indexOf`, `split`, `slice`, …) keep the legacy unified unbox; same dispatcher can be applied incrementally.
- A parallel `@.str.N.raw : i64` constant global per literal was not added: literal operands already reach proven-heap handles via `load` + `bitcast` + `and` (folds post-O3), with no per-literal binary growth.

## Summary by CodeRabbit

* **New Features**
  * Improved string operation performance through optimized handling of string values.
  * Faster `.length`, concatenation, equality, relational comparisons, and character-access methods.
  * Supports both short inline strings and longer strings while preserving existing behavior.
  * Improved string and number comparison handling, including Unicode and mixed representations.

* **Bug Fixes**
  * Preserved aliasing, `undefined` handling, Unicode correctness, and transitions from short to longer strings.
  * Added safeguards to prevent incorrect results during memory movement.

* **Tests**
  * Added comprehensive coverage for concatenation, comparisons, lengths, character access, and edge cases.

### CODE REVIEW

File: changelog.d/6909-repsel-p3a-canonical-str-locals.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+`SHORT_STRING_TAG` SSO) at rest. Storage, shadow-slot GC binding, and every
+alias/refcount demote stay exactly the pre-phase model (zero GC changes; SSO
+stays by-value); the rep is a compile-time proof the string-op lowerings
+consume to tag-dispatch inline instead of routing operands through
+`js_get_string_pointer_unified` (which heap-materializes SSO and
+number-coerces):
+
+- `s += rhs` self-append: both-heap → raw `js_string_append(h,h)` (keeps the
+  refcount==1 in-place path), both-string-SSO → `js_string_concat_box`, else
+  the exact legacy sequence (annotation lies degrade to today's behavior).
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply changelog.d/6909-repsel-p3a-canonical-str-locals.md]: Fixed in f569123f2 — now reads `js_string_append(lhs_h, rhs_h)`, and the arm description distinguishes the SSO-dest case explicitly.

File: crates/perry-runtime/src/string/compare.rs [hunk 2]
```diff
@@ -87,6 +87,67 @@ pub extern "C" fn js_string_equals(a: *const StringHeader, b: *const StringHeade
...
+    let mut b_num = None;
+    let a_view = bytes_of(a, &mut a_scratch, &mut a_num);
+    let b_view = bytes_of(b, &mut b_scratch, &mut b_num);
+    match (a_view, b_view) {
+        (None, None) => 0,
+        (None, Some(_)) => -1,
+        (Some(_), None) => 1,
+        (Some((a_ptr, a_len)), Some((b_ptr, b_len))) => unsafe {
+            let a_bytes = std::slice::from_raw_parts(a_ptr, a_len as usize);
+            let b_bytes = std::slice::from_raw_parts(b_ptr, b_len as usize);
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply crates/perry-runtime/src/string/compare.rs]: Valid — fixed in f569123f2. Restructured into two phases exactly as suggested: `number_bytes` (the only allocating step, `js_number_to_string` → owned `Vec<u8>`) completes for BOTH operands before `view_of` takes any heap payload pointer or SSO scratch view; nothing after phase 1 allocates, so the `from_raw_parts` reads can no longer observe a moved string.

Validation: unit test extended with number-vs-heap and number-vs-SSO mixes in both operand orders, and an end-to-end probe (20k iterations of `heapString > numberLie` / `numberLie < heapString` with a fresh heap string per iteration) runs clean under `PERRY_GC_FORCE_EVACUATE=1` and produces output identical to the flag-off (legacy `js_get_string_pointer_unified`) arm.

**Comment by proggeramlug** [comment]
Review nitpicks addressed in f569123f2:

- **`expr/compare.rs` duplicated heap-tag dispatch prologue** — extracted into `canonical_str_cmp_dispatch` (shared by the equality and relational arms; each keeps only its predicate tail).
- **Magic shifted-tag comparands** — now `crate::nanbox::STRING_TAG_TOP16_I64` / `SHORT_STRING_TAG_TOP16_I64`, asserted against the u64 tags in `tag_strings_match_u64_values`, and used across `lower_string_method.rs`, `expr/compare.rs`, and `expr/property_get.rs`.
- **`syntactic_str` name-only call arm** — receiver-ambiguous method names (`slice`, `concat`, `replace`, `charAt`, …) now also require a syntactically-string receiver, so `arr.slice()` no longer classifies as a string-producing write; only the universal ToString / number-formatting family (`toString`/`toFixed`/`toPrecision`/`toExponential`) stays name-only. (As noted, a misclassification here was a missed deopt, not a correctness break — the lowerings re-check the runtime tag — but the scan is now honest about under-approximating.)
- **`note_canonical_i32_local` naming** — renamed to `note_canonical_local` (definition + all call sites); it prints the rep and now serves `SlotRep::Str` too.
- **Truncated `repsel_context_allows_canonical_str` doc sentence** — completed ("…so the two phases can be A/B-tested independently").

Re-validated after the changes: shadow-slot hygiene suite (11/11), the Phase 3a gap test byte-exact vs node in all four arms (flag on/off × `PERRY_GC_FORCE_EVACUATE=1`), 9/9 string-focused gap smoke tests, and the structural kernel audit (hot arms unchanged: `+=` heap arm = direct `js_string_append`, compares = direct `js_string_equals`/`js_string_compare`, `.length`/receiver arms call-free; zero `js_get_string_pointer_unified` outside cold fallback arms and the entry-once alias demote).
